visibility/src/main/java/org/apache/accumulo/visibility/VisibilityExpression.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.visibility;
+
+import java.util.Set;
+
+/**
+ * An opaque type that contains a parsed visibility expression. When this type is constructed with
+ * {@link #parse(String)} and then used with
+ * {@link VisibilityArbiter#isVisible(VisibilityExpression)} it can be more efficient and avoid
+ * reparsing the expression.
+ *
+ * <p>
+ * For reviewers : this type is similar to ColumnVisibility. This interface and impl have goal of
+ * being immutable which differs from column visibility. ColumnVisibility leaks internal
+ * implementation details in its public API, this type does not.
+ *
+ * TODO needs better javadoc.
+ *
+ * Below is an example of using this API.
+ *
+ * <pre>
+ *     {@code
+ * var auth1 = VisibilityExpression.quote("CAT");
+ * var auth2 = VisibilityExpression.quote("🦕");
+ * var auth3 = VisibilityExpression.quote("🦖");
+ * var visExp = VisibilityExpression
+ *     .parse("(" + auth1 + "&" + auth3 + ")|(" + auth1 + "&" + auth2 + "&" + auth1 + ")");
+ * System.out.println(visExp.getExpression());
+ * System.out.println(visExp.normalize());
+ * System.out.println(visExp.getAuthorizations());
+ * }
+ * </pre>
+ *
+ * The above example will print the following.
+ *
+ * <pre>
+ * (CAT&amp;"🦖")|(CAT&amp;"🦕"&amp;CAT)
+ * ("🦕"&amp;CAT)|("🦖"&amp;CAT)
+ * [🦖, CAT, 🦕]
+ * </pre>
+ *
+ * @since ???
+ */
+// TODO could name VisibilityLabel
```

Comment:
What about `AccessExpression`? Looks like some terminology is already defined at https://accumulo.apache.org/docs/2.x/getting-started/glossary. We might be able to change the wording a little but still keep things relatively the same.

If we think in terms of ABAC (attribute based access control), then:

`Authorization` is an attribute
`Authorizations` is a set of attributes (typically associated with an entity)
`AccessExpression` is a boolean expression of attributes required to access an object (data / Key-Value)
`ColumnVisibility` is the part of the Key that holds the `AccessExpression`
`VisibilityArbiter` is the policy decision point that evaluates `Authorizations` against an `AccessExpression` to determine if the resource should be returned.

Reply:
Maybe `AccessPolicy` is a better term. I made the comments above before looking at https://nvlpubs.nist.gov/nistpubs/specialpublications/nist.sp.800-162.pdf, page 7.

Reply:
I like the names `AccessExpression` and `AccessPolicy`.  I don't like the name `VisibilityArbiter`, just chose something from a thesaurus as a place holder.

Reply:
@dlmarion I took a look at the document and did the following renames in 5c71441

 * renamed maven module from `accumulo-visibility` to `accumulo-access`
 * renamed package from `org.apache.accumulo.visibility` to `org.apache.accumulo.access`
 * renamed `VisibilityArbiter` to `AccessEvaluator`.  Looking at the document you linked this could also be an `AccessDecisionPoint`.
 * renamed `VisibilityExpression` to `AccessExpression`.  I like `AccessExpression` a bit more than `AccessPolicy`, the policy name seems really generic.  Thinking that `AccessExpression` is how Accumulo implements access policy.
 * renamed `IllegalVisibilityException` to `IllegalAccessExpressionException`

I like this much more than what was there and it feels like a big improvement to me.